tests/nn/beam_search_test.py

```diff
@@ -596,3 +630,202 @@ def test_length_normalized_sequence_log_prob_scorer(self):
         )
         expected_scores = expected_log_probs / length_normalization
         self._check_results(expected_top_k=expected_top_k, expected_log_probs=expected_scores)
+
+    def test_repeated_ngram_blocking_constraint_init_state(self):
+        ngram_size = 3
+        batch_size = 2
+        constraint = RepeatedNGramBlockingConstraint(ngram_size)
+
+        state = constraint.init_state(batch_size)
+        assert len(state) == batch_size
+        for beam_states in state:
+            assert len(beam_states) == 1
+            beam_state = beam_states[0]
+            assert len(beam_state.keys()) == 2
+            assert len(beam_state["current_prefix"]) == 0
+            assert len(beam_state["seen_ngrams"]) == 0
+
+    def test_repeated_ngram_blocking_constraint_apply(self):
+        ngram_size = 3
+        batch_size = 2
+        beam_size = 2
+        num_classes = 10
+        constraint = RepeatedNGramBlockingConstraint(ngram_size)
+
+        state = [
+            [
+                {"current_prefix": [0, 1], "seen_ngrams": {}},
+                {"current_prefix": [2, 3], "seen_ngrams": {(2, 3): [4]}},
+            ],
+            [
+                {"current_prefix": [4, 5], "seen_ngrams": {(8, 9): []}},
+                {"current_prefix": [6, 7], "seen_ngrams": {(6, 7): [0, 1, 2]}},
+            ],
+        ]
+        log_probabilities = torch.rand(batch_size, beam_size, num_classes)
+        constraint.apply(state, log_probabilities)
+
+        disallowed_locations = torch.nonzero(
+            log_probabilities == min_value_of_dtype(log_probabilities.dtype)
+        ).tolist()
+        assert len(disallowed_locations) == 4
+        assert [0, 1, 4] in disallowed_locations
+        assert [1, 1, 0] in disallowed_locations
+        assert [1, 1, 1] in disallowed_locations
+        assert [1, 1, 2] in disallowed_locations
+
+    def test_repeated_ngram_blocking_constraint_update_state(self):
+        ngram_size = 3
+        constraint = RepeatedNGramBlockingConstraint(ngram_size)
+
+        # We will have [2, 3] -> {5, 6} from batch index 0 and [4, 5] -> {0} and [6, 7] -> {3}
+        # from batch index
+        state = [
+            [
+                {"current_prefix": [0, 1], "seen_ngrams": {}},
+                {"current_prefix": [2, 3], "seen_ngrams": {(2, 3): [4]}},
+            ],
+            [
+                {"current_prefix": [4, 5], "seen_ngrams": {(8, 9): []}},
+                {"current_prefix": [6, 7], "seen_ngrams": {(6, 7): [0, 1, 2]}},
+            ],
+        ]
+        predictions = torch.LongTensor([[5, 6], [0, 3]])
+        backpointers = torch.LongTensor([[1, 1], [0, 1]])
```

Comment:
`backpointers` provided here to `update`, but not in `BeamSearch` (see my other comment).

Reply:
I would need to retrace the beam search that I used to write this test, but my guess is that this is not caught because the backpointer is always 0
https://github.com/allenai/allennlp/blob/5b1da908238794efb318da3327271cc945c6ba46/tests/nn/beam_search_test.py#L774